File: icfp2013/icfpc2013scala/src/test/scala/icfp/SimplificationTest.scala
package icfp

import org.junit.Test
import org.junit.Assert._
import icfp.parser.BVParsers

class SimplificationTest {
  
  @Test
  def testSimplifyShr() {
    val e = new BVParsers().parseExpression("(shr4 (shr1 x))")
    assertEquals("(shr5 x)", e.simplify.get.toString());
  }
  
  @Test
  def testSimplifyNotXor() {
    val e = new BVParsers().parseExpression("(not (xor 1 x))")
    assertEquals("(xor fffffffffffffffe x)", e.simplify.get.toString());
  }
  
  @Test
  def testSimplifyIf0() {
    val e = new BVParsers().parseExpression("(if0 x 1 1)")
    assertEquals(None, e.simplify);
  }
  
  @Test
  def testSimplifyWithMaxValue {
    val e = new BVParsers().parseExpression("(shr16 (and x (plus 1 1)))")
    assertEquals(None, e.simplify);
  }

}

File: icfp2013/icfpc2013java/src/test/java/search/SimplificationTest.java
package search;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

import domain.Identifier;
import domain.expression.IdExpression;
import domain.expression.LiteralExpression;
import domain.expression.ShrExpression;

public class SimplificationTest {
	@Test
	public void testSimplifyShr() {
		IdExpression id = new IdExpression(new Identifier("x"));
		ShrExpression e = new ShrExpression(new ShrExpression(id, 4), 1);
		assertEquals(new ShrExpression(id, 5), e.simplify());
		assertEquals(e, e.simplify().getCanonicalForm());
	}
	
	@Test
	public void testSimplifyShrConstant() {
		ShrExpression e = new ShrExpression(new ShrExpression(LiteralExpression.ONE, 4), 1);
		assertEquals("(shr1 (shr4 1))", e.simplify().toString());
		assertEquals(e, e.simplify().getCanonicalForm());
	}
}
